plonky3/src/circuit_builder.rs

```diff
@@ -6,28 +6,42 @@
 //! everywhere save for at row j is constructed to constrain s * (pub - x) on
 //! every row.
 
-use std::collections::{BTreeMap, HashSet};
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, BTreeSet, HashMap},
+    sync::Mutex,
+};
 
 use crate::params::{Commitment, FieldElementMap, Plonky3Field, ProverData};
 use p3_air::{Air, AirBuilder, AirBuilderWithPublicValues, BaseAir, PairBuilder};
 use p3_matrix::{dense::RowMajorMatrix, Matrix};
 use powdr_ast::analyzed::{
     AlgebraicBinaryOperation, AlgebraicBinaryOperator, AlgebraicExpression,
-    AlgebraicUnaryOperation, AlgebraicUnaryOperator, Analyzed, Identity, IdentityKind,
-    PolynomialType, SelectedExpressions,
+    AlgebraicUnaryOperation, AlgebraicUnaryOperator, Analyzed, Challenge, Identity, IdentityKind,
+    PolyID, PolynomialType, SelectedExpressions, SymbolKind,
 };
+
+use p3_uni_stark::{CallbackResult, MultiStageAir, MultistageAirBuilder, NextStageTraceCallback};
+use powdr_ast::parsed::visitor::ExpressionVisitable;
+
 use powdr_executor::witgen::WitgenCallback;
-use powdr_number::{DegreeType, FieldElement};
+use powdr_number::FieldElement;
+
+type Witness<T> = Mutex<RefCell<Vec<(String, Vec<T>)>>>;
```

Comment:
That would not be enough because in `stark.rs` when calling `prove_with_key` we pass the circuit twice (once as air and once as `NextStageTraceCallback`) so this only works with immutable references.

That said there may well be simplifications, but I think they will be easier to apply once everything is in the same codebase.